frontend/src/app/models/ngram.ts

```diff
@@ -0,0 +1,64 @@
+import { Params } from '@angular/router';
+import * as _ from 'lodash';
+
+import { StoreSync } from '../store/store-sync';
+import { Store } from '../store/types';
+
+export interface NgramSettings {
+    size: number;
+    positions: string;
+    freqCompensation: boolean;
+    analysis: string;
+    maxDocuments: number;
+    numberOfNgrams: number;
+    dateField: string;
+}
+
+export class NgramParameters extends StoreSync<NgramSettings> {
+
+    keysInStore = ['ngramSettings'];
+
+    constructor(store: Store) {
+        super(store);
+        this.connectToStore();
+    }
+
+    stringifyNgramSettings(state: NgramSettings): string {
+        return [`s:${state.size}`,`p:${state.positions}`,`c:${state.freqCompensation}`,
+            `a:${state.analysis}`,`m:${state.maxDocuments}`,`n:${state.numberOfNgrams}`,
+            `f:${state.dateField}`].join(',')
+    }
+
+    stateToStore(state: NgramSettings): Params {
+        return { ngramSettings: this.stringifyNgramSettings(state)}
+    }
+
+    storeToState(params: Params): NgramSettings {
+        if (_.has(params, 'ngramSettings')) {
+            const stringComponents = params['ngramSettings'].split(',');
+            return {
+                size: parseInt(this.findSetting('s', stringComponents), 10),
+                positions: this.findSetting('p', stringComponents),
+                freqCompensation: this.findSetting('c', stringComponents) === 'true',
+                analysis: this.findSetting('a', stringComponents),
+                maxDocuments: parseInt(this.findSetting('m', stringComponents), 10),
+                numberOfNgrams: parseInt(this.findSetting('n', stringComponents), 10),
+                dateField: this.findSetting('f', stringComponents)
+            }
+        }
+        return {
+            size: 2,
+            positions: 'any',
+            freqCompensation: false,
+            analysis: 'none',
+            maxDocuments: 50,
+            numberOfNgrams: 10,
+            dateField: 'date'
```

Comment:
The way the date field is handled looks a bit shaky, since the corpus may not actually have a field with this name. I think this was already an issue, though. Have you checked that the visualisation still works with corpora like parliament-denmark-old?

Reply:
Yes, good point. The problem is that when the `ngramParameters` model is initialized, the `corpus` input hasn't been set yet, so we cannot set the date field. Perhaps it would break functionality least if I excluded the date field from the route parameters for now. This is exactly why I was wondering how realistic it is to transfer data conversion logic to models, as we will also have to transfer knowledge about possible states in the component to the model.

Reply:
Hmm, I tried and abandoned the idea to initialize `ngramParameters` in `ngOnChanges`. That seems to work in practice, but unit testing depending on change detection is a bit of a nuisance. I decided to remove the date field from the routing parameters for now: the number of corpora in which different date fields are available is so limited that I think it's not too important to include the date field in the route at this moment.